test/smoother/test_smoother.cpp

```diff
@@ -0,0 +1,330 @@
+    model.set_threads(1);
+
+    OutputUpdater output_updater(model.device);
+
+    //////////////////////////////////////////////////////////////////////
+    // Training
+    //////////////////////////////////////////////////////////////////////
+    unsigned seed = 0;
+    std::default_random_engine seed_e(seed);
+    int n_epochs = 50;
+    int batch_size = 1;
+    float sigma_obs = 1.0;
+
+    int iters = train_db.num_data / batch_size;
+    std::cout << "num_iter: " << iters << "\n";
+    std::vector<float> x_batch(batch_size * train_db.nx, 0.0f);
+    std::vector<float> var_obs(batch_size * train_db.ny, pow(sigma_obs, 2));
+    std::vector<float> y_batch(batch_size * train_db.ny, 0.0f);
+    std::vector<int> batch_idx(batch_size);
+    std::vector<float> mu_a_output(batch_size * train_db.ny, 0);
+    std::vector<float> var_a_output(batch_size * train_db.ny, 0);
+    auto data_idx = create_range(train_db.num_data);
+    float decay_factor = 0.95f;
+    float min_sigma_obs = 0.3f;
+    // for smoother
+    std::vector<float> mu_zo_smooths(0.0f, input_seq_len),
+        var_zo_smooths(0.0f, input_seq_len);
+    std::vector<float> mu_sequence(input_seq_len, 1.0f);
+    float mu_zo = 0;
+    // Number of observations before training time to be inferred. These
+    // obervations are 0 in training data.
+    int infer_window_len = 48;
+
+    for (int e = 0; e < n_epochs; e++) {
+        if (e > 0) {
+            // Decay observation noise
+            decay_obs_noise(sigma_obs, decay_factor, min_sigma_obs);
+            std::vector<float> var_obs(batch_size * train_db.ny,
+                                       pow(sigma_obs, 1));
+        }
+        std::cout << "################\n";
+        std::cout << "Epoch #" << e + 1 << "/" << n_epochs << "\n";
+        std::cout << "Training...\n";
+        auto start = std::chrono::steady_clock::now();
+        for (int i = 0; i < iters; i++) {
+            // Load data
+            get_batch_idx(data_idx, i * batch_size, batch_size, batch_idx);
+            get_batch_data(train_db.x, batch_idx, train_db.nx, x_batch);
+            get_batch_data(train_db.y, batch_idx, train_db.ny, y_batch);
+
+            //  replace nan in input x by the lstm_prediction
+            if (i < input_seq_len + infer_window_len) {
+                replace_with_prediction(mu_sequence, x_batch);
+            }
+
+            // Forward
+            model.forward(x_batch);
+            output_updater.update(*model.output_z_buffer, y_batch, var_obs,
+                                  *model.input_delta_z_buffer);
+
+            // Backward pass
+            model.backward();
+            model.step();
+
+            // Output the prediction
+            mu_zo = model.output_z_buffer->mu_a[0];
+            // Add new prediction to mu_sequence
+            add_prediction_to_sequence(input_seq_len, mu_zo, mu_sequence);
+        }
+
+        // Smoother
+        auto [mu_zo_smooths, var_zo_smooths] = model.smoother();
```

Comment:
It causes a warning while compiling. Currently we use c++14 and this feature `auto [...]` is only available on C++17 
```shell
warning: structured bindings only available with ‘-std=c++17’ or ‘-std=gnu++17’ [-Wc++17-extensions]
[build]   223 |         auto [mu_zo_smooths, var_zo_smooths] = model.smoother();
```
Here is the link to an [example](https://github.com/lhnguyen102/cuTAGI/blob/main/test/fnn/test_fnn_mnist_cpu.cpp#L214) how you can handle this situation.

Reply:
revised following the suggestion.

Reply:
I did not see the changes. The warning is still there. Am I missing something?

```shell
warning: structured bindings only available with ‘-std=c++17’ or ‘-std=gnu++17’ [-Wc++17-extensions]
[build]   223 |         auto [mu_zo_smooths, var_zo_smooths] = model.smoother();

```

Reply:
sorry. I forgot to commit, just committed.